main/actions/src/main/scala/sbt/compiler/Eval.scala

```diff
@@ -1,75 +1,163 @@
+ */
 final class EvalException(msg: String) extends RuntimeException(msg)
-// not thread safe, since it reuses a Global instance
-final class Eval(optionsNoncp: Seq[String], classpath: Seq[File], mkReporter: Settings => Reporter, backing: Option[File]) {
-  def this(mkReporter: Settings => Reporter, backing: Option[File]) = this(Nil, IO.classLocationFile[Product] :: Nil, mkReporter, backing)
-  def this() = this(s => new ConsoleReporter(s), None)
 
-  backing.foreach(IO.createDirectory)
-  val classpathString = Path.makeString(classpath ++ backing.toList)
-  val options = "-cp" +: classpathString +: optionsNoncp
+/**
+ * Represents the evaluation logic of sbt files.
+ *
+ * This method is not thread-safe since it reuses the same [[Global]] instance.
+ *
+ * @param optionsWithoutClasspath Scalac options that don't include 'cp'.
+ * @param classpath The classpath to compile and evaluate sbt code.
+ * @param mkReporter A function that turns settings into a reporter.
+ * @param target The directory where the compiled class files are saved.
+ */
+final class Eval(optionsWithoutClasspath: Seq[String],
+  classpath: Seq[File],
+  mkReporter: Settings => Reporter,
+  target: Option[File])
+    extends SbtParser
+    with SbtDefinitions
+    with SbtCodegen
+    with SbtHashGen {
+
+  def this(mkReporter: Settings => Reporter, target: Option[File]) =
+    this(Nil, IO.classLocationFile[Product] :: Nil, mkReporter, target)
+
+  def this() = this(s => new ConsoleReporter(s), None)
 
-  lazy val settings =
-    {
-      val s = new Settings(println)
-      val command = new CompilerCommand(options.toList, s)
-      s
-    }
-  lazy val reporter = mkReporter(settings)
   /**
-   * Subclass of Global which allows us to mutate currentRun from outside.
-   * See for rationale https://issues.scala-lang.org/browse/SI-8794
+   * Represent a subclass of [[Global]] to mutate the current run.
+   *
+   * This is necessary to control the Scalac pipeline.
    */
-  final class EvalGlobal(settings: Settings, reporter: Reporter) extends Global(settings, reporter) {
+  final class EvalGlobal(settings: Settings, reporter: Reporter)
+      extends Global(settings, reporter) {
     override def currentRun: Run = curRun
     var curRun: Run = null
   }
+
+  /* *********************************************************************** */
+  /* ******** Initialize the target, settings, reporter and global ********* */
+  /* *********************************************************************** */
+
+  // TODO: provide a way to cleanup target directory
+  target.foreach(IO.createDirectory)
+
+  val scalacOptions: Seq[String] = {
+    val completeClasspath = Path.makeString(classpath ++ target.toList)
+    "-cp" +: completeClasspath +: optionsWithoutClasspath
+  }
+
+  lazy val settings: Settings = {
+    val settings0 = new Settings(println)
+    // Don't touch this line if you like order in the universe
+    new CompilerCommand(scalacOptions.toList, settings0)
```

Comment:
The reason for this is that compiler command mutates global (IIRC). Even if we don't use it, we need it to evaluate.